### ISSUE TITLE
`opt+cmd+q` not latex-sensitive

### ISSUE THREAD

**Issue by daviwil** [atom/atom#18029]
_From @rothos on November 7, 2014 17:45_

I’m new to Atom and am not sure if this issue belongs here, so let me know if it belongs somewhere else, e.g. in the issues for the package [area/language-latex](http://github.com/area/language-latex), which handles LaTeX language compatibility.

The issue #12 points out that Autoflow does not respect LaTeX comments, but really the problem is more broad. Here are a few misbehaviors of Autoflow in LaTeX. (Here I am defining the “correct” behavior as that of Sublime Text’s [LaTeXTools](https://github.com/SublimeText/LaTeXTools) package, which is what I’m used to.)
1. **Autoflow does not respect that lines beginning with a LaTeX command usually should not be reformatted.**
   
   Input (the pipe represents where my cursor when I hit `alt+cmd+q`; no text is highlighted):
   
   ```
   \section{Introduction}
   |This is the
   introduction paragraph.
   ```
   
   Expected output:
   
   ```
   \section{Introduction}
   This is the introduction paragraph.
   ```
   
   Actual output:
   
   ```
   \section{Introduction} This is the introduction paragraph.
   ```
   
   I think the rule that Sublime’s LaTeXTools uses is that if a line begins with `\` it is not reformatted.
2. **Autoflow does not fully respect LaTeX comments (I).** This is pointed out in #12. Although comments are not treated as normal text like the `\section` command is in the first example, Autoflow does not correctly insert a symbol `%` upon reformatting.
   
   Input (the pipe represents where my cursor when I hit `alt+cmd+q`; no text is highlighted):
   
   ```
   \section{Introduction}
   |% In this section the subject is introduced. Lorem ipsum dolor sit amet, adipiscing elit.
   ```
   
   Expected output:
   
   ```
   \section{Introduction}
   % In this section the subject is introduced. Lorem ipsum dolor sit amet,
   % adipiscing elit.
   ```
   
   Actual output:
   
   ```
   \section{Introduction}
   % In this section the subject is introduced. Lorem ipsum dolor sit amet,
   adipiscing elit.
   ```
3. **Autoflow does not fully respect LaTeX comments (II).** Similarly to above, Autoflow does not remove a `%` when it merges two comment lines.
   
   Input (the pipe represents where my cursor when I hit `alt+cmd+q`; no text is highlighted):
   
   ```
   \section{Introduction}
   |% In this section
   % the subject is introduced.
   ```
   
   Expected output:
   
   ```
   \section{Introduction}
   % In this section the subject is introduced.
   ```
   
   Actual output:
   
   ```
   \section{Introduction}
   % In this section % the subject is introduced.
   ```

_Copied from original issue: atom/autoflow#18_

**Comment by daviwil** [comment]
_From @thmosqueiro on December 6, 2014 0:10_

I've observed a similar behavior with **item 1** while testing it with atomic emacs. I believe _it's not related to latex_ only.

For instance, I've tried this with python.

```
def foo(a,b):
    """ Doc...
    """
    return
```

is reflowed to

```
def foo(a,b):  """ Doc...
    """
    return
```

Both problems seem to have the same root: identifying the previous line.

However, I could not reproduce the issues related with **items 2** and **3**. Comments are properly reflowed.

**Comment by daviwil** [comment]
_From @JamesGuthrie on September 17, 2015 13:27_

It appears to me as though autoflow's paragraph reflowing functionality is quite broken with latex, independent of its interaction with **item 1** above.

The following text:

```
This text illustrates
atom's incorrect paragraph reflowing functionality. While long lines are correctly broken into shorter lines,
short lines within a paragraph
are not reflowed into the lines above them in order to
produce one coherent paragraph.
```

when reflowed in a latex document formats to:

```
This text illustrates
 atom's incorrect paragraph reflowing functionality. While
long lines are correctly broken into shorter lines,
 short lines within a
paragraph
 are not reflowed into the lines above them in order to
 produce one
coherent paragraph.
```

Whereas in a new, blank document, reflowing the above results in:

```
This text illustrates atom's incorrect paragraph reflowing functionality. While
long lines are correctly broken into shorter lines, short lines within a
paragraph are not reflowed into the lines above them in order to produce one
coherent paragraph.
```

I'm using atom 1.0.11.

**Comment by daviwil** [comment]
_From @JamesGuthrie on September 18, 2015 8:57_

A follow-up: it looks as though the root my problem is that the document I was editing uses CRLF line endings. While autoflow "correctly" reflowed the text, it only removes \n, not \r from the line endings. The result is that the reflowed text is broken on the new line breaks that autoflow adds, and the \r component of the linebreak which was not removed from the original line breaks.

**Comment by daviwil** [comment]
_From @JamesGuthrie on September 18, 2015 12:47_

Another follow up, my issue should be fixed by #35
